### PR TITLE
Resize Point and Rectangle controls based on overall width.

### PR DESCRIPTION
**Pull request by CartBlanche**
Controls resized as per control width

Supposed to fix this - https://github.com/xamarin/Xamarin.PropertyEditing/issues/518

Narrow:
screen shot 2019-02-14 at 10 06 13

Wide:
screen shot 2019-02-14 at 10 04 12

### CODE REVIEW

**Comment by CartBlanche** [comment]
@vancura Could you please take a look at the screenshot above.

**Comment by ermau** [comment]
What does this look like when it's shrunk down small? Seems like we need a max size on the spacing rather than a fixed size.

**Comment by CartBlanche** [comment]
@ermau There are 2 screenshot above. Once with it shrunk quite small and another larger one. Do you mean even smaller than that? We can reduce the spacing, but that will only gain us 50-100pixels.

**Comment by ermau** [comment]
If the top image is representative of how this PR behaves, I definitely think that it should line up with the other controls like the second screenshot.

**Comment by vancura** [comment]
No, @CartBlanche, sorry, this is not correct. See:

image

**Comment by CartBlanche** [comment]
@vancura OK

**Comment by CartBlanche** [comment]
@vancura Can you comment on the spacing between X & Y controls. Should the spacing be less now, between them? As in Right of X control, to Left of Y control.

**Comment by CartBlanche** [comment]
Reduced spacing with narrowest width...
screen shot 2019-02-08 at 16 13 14

Still reduced spacing with wider width...
screen shot 2019-02-08 at 16 13 40

**Comment by vancura** [comment]
It should be `18pt`:

image

Your gap seems to be `32pt`.

**Comment by vancura** [comment]
Actually, sorry, can you make it `10pt`? So it matches other alignments.

**Comment by CartBlanche** [comment]
@vancura Here is is a 10, according to Constraints...

screen shot 2019-02-11 at 13 53 20

**Comment by vancura** [comment]
This is great, thanks!

BTW the right column is still here, but I think you mentioned you would fix it in another PR? Sorry, my memory is sparse.

**Comment by CartBlanche** [comment]
@vancura Yes, right column removal will in in another PR.

**Comment by vancura** [comment]
Is this the ticket where I can ask you to add two label tweaks?

Your version: 
image

My version:
image

Differences:
- [ ] Your label font size is `10pt`, make it `9pt`.
- [ ] Move the label `1pt` up (make the vertical gap between the textfield and label `1pt` smaller) – this might be unnecessary as the label font size might fix this, so let's see.

Thanks!

**Comment by CartBlanche** [comment]
@vancura 9pt Font and 1 pt up. Any good?

screen shot 2019-02-11 at 15 14 40

**Comment by CartBlanche** [comment]
@vancura To my view of symmetry, the right of the `Property Filter` search box, should really line up with the right edge of the controls, rather than the property button. It looks lop sided to me. Could just be me of course and that might be a tweak for another PR, if it is needed.

**Comment by vancura** [comment]
@CartBlanche In fact, the whole header part was designed years ago, but never implemented. It's created to be globally used, not just in Property Pads.

The alignment will make more sense when the gap on the right is correct, I promise :)

**Comment by CartBlanche** [comment]
@ermau if you have a minute look over the code changes. Should hopefully be straight forward.

**Comment by ermau** [comment]
@vancura Those other measurements are unrelated to this PR, I'm assuming they're logged in the devops set of issues.

File: Xamarin.PropertyEditing.Mac/Controls/BasePointEditorControl.cs [hunk 1]
```diff
@@ -42,10 +48,19 @@ protected BasePointEditorControl (IHostResourceProvider hostResources)
 			AddSubview (YEditor);
 
 			this.AddConstraints (new[] {
-				NSLayoutConstraint.Create (XEditor, NSLayoutAttribute.Width, NSLayoutRelation.Equal, 1f, 90f),
+				NSLayoutConstraint.Create (XLabel, NSLayoutAttribute.Top, NSLayoutRelation.Equal, this,  NSLayoutAttribute.Top, 1f, 16f),
```
Comment (ermau) [inline Xamarin.PropertyEditing.Mac/Controls/BasePointEditorControl.cs]: Shouldn't the labels constrain relative to what they're labeling? 

`NSLayoutConstraint.Create (XLabel, NSLayoutAttribute.Top, NSLayoutRelation.Equal, XEditor,  NSLayoutAttribute.Bottom, 1f, 2),`

File: Xamarin.PropertyEditing.Mac/Controls/RectangleEditorControl.cs [hunk 2]
```diff
@@ -14,29 +14,28 @@ protected RectangleEditorControl (IHostResourceProvider hostResources)
...
+			
 			HeightLabel.StringValue = "HEIGHT";
 
 			HeightEditor.Frame = new CGRect (132, 13, 90, 20);
+
+			this.AddConstraints (new[] {
+				NSLayoutConstraint.Create (XLabel, NSLayoutAttribute.Left, NSLayoutRelation.Equal, XEditor, NSLayoutAttribute.CenterX, 1f, -10f),
+				NSLayoutConstraint.Create (YLabel, NSLayoutAttribute.Left, NSLayoutRelation.Equal, YEditor, NSLayoutAttribute.CenterX, 1f, -10f),
+				NSLayoutConstraint.Create (WidthLabel, NSLayoutAttribute.Left, NSLayoutRelation.Equal, WidthEditor, NSLayoutAttribute.CenterX, 1f, -25f),
+				NSLayoutConstraint.Create (HeightLabel, NSLayoutAttribute.Left, NSLayoutRelation.Equal, HeightEditor, NSLayoutAttribute.CenterX, 1f, -25f),
```
Comment (ermau) [inline Xamarin.PropertyEditing.Mac/Controls/RectangleEditorControl.cs]: You can specify `CenterX` on the `HeightLabel` so you don't need to do any manual adjustment that will assuredly break when the labels are allowed to be changed.

`NSLayoutConstraint.Create (HeightLabel, NSLayoutAttribute.CenterX, NSLayoutRelation.Equal, HeightEditor, NSLayoutAttribute.CenterX, 1f, 0),`

File: Xamarin.PropertyEditing.Mac/Controls/BasePointEditorControl.cs (hunk 1 above)
Comment (ermau) [reply Xamarin.PropertyEditing.Mac/Controls/BasePointEditorControl.cs]: Also why does this happen here when the other label constraints are added in the base?

File: Xamarin.PropertyEditing.Mac/Controls/BasePointEditorControl.cs [hunk 3]
```diff
@@ -42,10 +48,19 @@ protected BasePointEditorControl (IHostResourceProvider hostResources)
...
+				NSLayoutConstraint.Create (XLabel, NSLayoutAttribute.Height, NSLayoutRelation.Equal, 1f, DefaultControlHeight),
+
+				NSLayoutConstraint.Create (XEditor, NSLayoutAttribute.Left, NSLayoutRelation.Equal, this,  NSLayoutAttribute.Left, 1f, 0f),
+				NSLayoutConstraint.Create (XEditor, NSLayoutAttribute.Right, NSLayoutRelation.Equal, YEditor,  NSLayoutAttribute.Left, 1f, -10f),
+				NSLayoutConstraint.Create (XEditor, NSLayoutAttribute.Width, NSLayoutRelation.Equal, this,  NSLayoutAttribute.Width, 1f, -280f),
```
Comment (ermau) [inline Xamarin.PropertyEditing.Mac/Controls/BasePointEditorControl.cs]: This seems like a wild guess and produces unusable results. Use the ratio argument to say "half the host, minus 10 for spacing" :
`NSLayoutConstraint.Create (XEditor, NSLayoutAttribute.Width, NSLayoutRelation.Equal, this,  NSLayoutAttribute.Width, 0.5f, -10f),`

**Comment by ermau** [comment]
image

On first load, the sizing of these controls isn't at all related with its host width. I think this will probably be addressed by some of the changes I suggested.

File: Xamarin.PropertyEditing.Mac/Controls/RectangleEditorControl.cs (hunk 2 above)
Comment (CartBlanche) [reply Xamarin.PropertyEditing.Mac/Controls/RectangleEditorControl.cs]: No, because it will centre on the whole control, including spinners, which we don't want it to do, we just want it centred on the editing part of the control. At least that was the initial design from @vancura . If that has changed, then I'll change it to be centre on everything.

File: Xamarin.PropertyEditing.Mac/Controls/BasePointEditorControl.cs (hunk 1 above)
Comment (CartBlanche) [reply Xamarin.PropertyEditing.Mac/Controls/BasePointEditorControl.cs]: IIRC this caused a constraints crash/collision, but I'll double check it.

File: Xamarin.PropertyEditing.Mac/Controls/RectangleEditorControl.cs (hunk 2 above)
Comment (ermau) [reply Xamarin.PropertyEditing.Mac/Controls/RectangleEditorControl.cs]: You could include a manual offset for the spinners themselves, a negative offset of half the width of the spinners.

**Comment by CartBlanche** [comment]
My first load, looks fine. Launch, click `Mock 2`. Give me this...

screen shot 2019-02-13 at 19 12 13

What are the steps to repro yours?

File: Xamarin.PropertyEditing.Mac/Controls/Layout.cs [hunk 4]
```diff
@@ -3,6 +3,6 @@ namespace Xamarin.PropertyEditing.Mac
 {
 	internal static class Layout
 	{
-		internal const float GoldenRatioLeft = .38196601125f;
```
Comment (ermau) [inline Xamarin.PropertyEditing.Mac/Controls/Layout.cs]: Why was this changed? This will affect a number of things and I'm not sure how you arrived at the new number, but I'm pretty sure it's incorrect.

**Comment by ermau** [comment]
> My first load, looks fine. Launch, click Mock 2. Give me this...

It was reliable, now it's sporadic. I'll keep poking at it.

File: Xamarin.PropertyEditing.Mac/Controls/Layout.cs (hunk 4 above)
Comment (CartBlanche) [reply Xamarin.PropertyEditing.Mac/Controls/Layout.cs]: I was playing around with the left margin and forget to reset it back. Will be fixed in next push.

File: Xamarin.PropertyEditing.Mac/Controls/RectangleEditorControl.cs (hunk 2 above)
Comment (CartBlanche) [reply Xamarin.PropertyEditing.Mac/Controls/RectangleEditorControl.cs]: The Labels were hard coded as `X`, `Y`, `WIDTH`, `HEIGHT` etc because we had a discussion about how they wouldn't change. Otherwise I would have put them in for localisation resources. We discussed this with Alan like a year or possibly more ago.

File: Xamarin.PropertyEditing.Mac/Controls/BasePointEditorControl.cs (hunk 1 above)
Comment (CartBlanche) [reply Xamarin.PropertyEditing.Mac/Controls/BasePointEditorControl.cs]: Fixed.

File: Xamarin.PropertyEditing.Mac/Controls/BasePointEditorControl.cs (hunk 3 above)
Comment (CartBlanche) [reply Xamarin.PropertyEditing.Mac/Controls/BasePointEditorControl.cs]: That width shouldn't be there. Removed.

**Comment by CartBlanche** [comment]
@ermau Please grab latest.

File: Xamarin.PropertyEditing.Mac/Controls/RectangleEditorControl.cs (hunk 2 above)
Comment (ermau) [reply Xamarin.PropertyEditing.Mac/Controls/RectangleEditorControl.cs]: We discussed them not being localized which is not the same as not being changed, they are a property name and therefore subject to the platform. We haven't implemented the functionality yet, but they will absolutely have the option of changing in the future. Beyond that, your manual offset is still subject to the font and font size. Is there another technical concern with doing it this way?

**Comment by CartBlanche** [comment]
@vancura Do the horizontal alignment on the Point/Rect etc control looks fine to you...
screen shot 2019-02-15 at 16 25 39

File: Xamarin.PropertyEditing.Mac/Controls/RectangleEditorControl.cs (hunk 2 above)
Comment (CartBlanche) [reply Xamarin.PropertyEditing.Mac/Controls/RectangleEditorControl.cs]: I think calculating the offsets, with font info will incur an overhead, but I've made the change. We'll find out if that is likely or not. Just awaiting Vaclav's approval that it looks fine for him.

Comment (ermau) [reply Xamarin.PropertyEditing.Mac/Controls/RectangleEditorControl.cs]: You're right, that would incur an overhead. Was there an issue with my suggested solution?

Comment (CartBlanche) [reply Xamarin.PropertyEditing.Mac/Controls/RectangleEditorControl.cs]: @ermau  Do you mean this...
> You could include a manual offset for the spinners themselves, a negative offset of half the width of the spinners.

?

Comment (ermau) [reply Xamarin.PropertyEditing.Mac/Controls/RectangleEditorControl.cs]: > Do you mean this?

That and
`NSLayoutConstraint.Create (HeightLabel, NSLayoutAttribute.CenterX, NSLayoutRelation.Equal, HeightEditor, NSLayoutAttribute.CenterX, 1f, 0),`

Comment (CartBlanche) [reply Xamarin.PropertyEditing.Mac/Controls/RectangleEditorControl.cs]: `NSLayoutConstraint.Create (HeightLabel, NSLayoutAttribute.CenterX, NSLayoutRelation.Equal, HeightEditor, NSLayoutAttribute.CenterX, 1f, 0),` will not take into account the width of the actual text in the label, it will position it exactly in the centre of the label. Adding the spinner offsets, unfortunately isn't enough for it to position the text correctly. 

As allowing custom labels in on the road-map, doing the width calculations is the best way to ensure we handle things correctly, regardless of font and the text users might use. Please take a look at the calculations. There might be another way to minimise the calculations overheads. It doesn't look massive right now, but might be worth another pair of eyes.

Comment (ermau) [reply Xamarin.PropertyEditing.Mac/Controls/RectangleEditorControl.cs]: > will not take into account the width of the actual text in the label, it will position it exactly in the centre of the label.

The label already sizes to text, how would the center of the label not also be the center of the text?
image

I checked out your branch and tested this change to the thickness editor (we can actually skip the spinner offset by constraint directly to the textbox):
```csharp
this.AddConstraints (new[] {
	NSLayoutConstraint.Create (XLabel, NSLayoutAttribute.CenterX, NSLayoutRelation.Equal, XEditor.Subviews[0], NSLayoutAttribute.CenterX, 1f, 0),
	NSLayoutConstraint.Create (YLabel, NSLayoutAttribute.CenterX, NSLayoutRelation.Equal, YEditor.Subviews[0], NSLayoutAttribute.CenterX, 1f, 0),
	NSLayoutConstraint.Create (WidthLabel, NSLayoutAttribute.CenterX, NSLayoutRelation.Equal, WidthEditor.Subviews[0], NSLayoutAttribute.CenterX, 1f, 0),
	NSLayoutConstraint.Create (HeightLabel, NSLayoutAttribute.CenterX, NSLayoutRelation.Equal, HeightEditor.Subviews[0], NSLayoutAttribute.CenterX, 1f, 0),
});
```

image

As you can measure, the thickness labels are centered perfectly (give or take 1px because the widths aren't even) on the textbox above it.

Comment (CartBlanche) [reply Xamarin.PropertyEditing.Mac/Controls/RectangleEditorControl.cs]: I hadn't thought of using `SubViews[0]`. That works really well and simplifies the code across the board and no calcs on our part.

Comment (CartBlanche) [reply Xamarin.PropertyEditing.Mac/Controls/RectangleEditorControl.cs]: Done.
